docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,449 @@
+    # file will be uploaded to MEDIA_ROOT/user_<id>/<filename>
+    return "user_{0}/{1}".format(instance.user.id, filename)
+
+  class MyModel(models.Model):
+    upload = models.FileField(upload_to=user_directory_path) # <-- the 'upload_to' parameter defines our custom function
+
+Note that this source is already known by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sourceModel
+      data:
+        - [
+            "django.db.models.FileField!",
+            "Call.Argument[upload_to:].Parameter[1,filenam:]",
+            "remote",
+          ]
+
+
+- Since we're adding a new taint source, we add a tuple to the **sourceModel** extensible predicate.
+- The first column, **"django.db.models.FileField!"**, is a dotted path to the **FileField** class from the **django.db.models** package.
+  The **!** at the end of the type name indicates that we are looking for the class itself rather than instances of this class.
+
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+  
+    - **Call** selects calls to the class. That is, constructor calls.
+    - **Argument[upload_to:]** selects the named argument named **upload_to**. Note that the colon at the end of the argument name indicates that we are looking for a named argument.
+    - **Parameter[1,filename:]** selects parameters of the callback function. We provide two parameters to search for, namely the second positional parameter and the named parameter **filename**.
+
+- Finally, the kind **remote** indicates that this is considered a source of remote flow.
+
+Example: Adding flow through 're.compile'
+----------------------------------------------
+
+In this example, we'll show how to add flow through calls to `re.compile`.
+`re.compile` returns a compiled regular expression for efficient evaluation, but the pattern to be compiled is stored in the `pattern` attribute of the resulting object.
+
+.. code-block:: python
+
+  import re
+
+  let y = re.compile(pattern = x); // add value flow from 'x' to 'y.pattern'
+
+Note that this flow is already recognized by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: summaryModel
+      data:
+        - [
+            "re",
+            "Member[compile]",
+            "Argument[0,pattern:]",
+            "ReturnValue.Attribute[pattern]",
+            "value",
+          ]
+
+
+- Since we're adding flow through a function call, we add a tuple to the **summaryModel** extensible predicate.
+- The first column, **"re"**, begins the search for relevant calls at places where the **re** package is imported.
```

Comment:
```suggestion
- The first column, **re**, begins the search for relevant calls at places where the **re** package is imported.
```
Generally, you've left off the quotes when referring to bits of the MaD rows, so maybe the same should be the case here? (There are other instances elsewhere in the document.)

Reply:
Hm, I think that I started having full column values with their quotes and parts without, but even this is not followed consistently. Let us simply remove all quotes, the examples show where quotes should be put...